docs/roadmap.md

```diff
@@ -0,0 +1,83 @@
+# Roadmap
+
+This is a list of goals we have. Many of them are quite independent.
+
+> **Note:** Most people contributing to Jujutsu do so in their spare time, which
+>  means that we cannot attach any target dates to any of the goals below.
+
+## Support for copies and renames
+
+We want to support copy tracing in a way that leaves it up to the commit backend
+to either record or detect copies. That should let us work with existing Git
+repos (Git does not record copies, it detects them on the fly) as well as with
+very large repos where detection would be too slow. See
+[design doc][copy-design-doc].
+
+## Forge integrations
+
+We would like to make it easier to work with various popular forges by providing
```

Comment:
Am I hearing "dynamically loadable extensions"? :) 'cause that sounds much better than trying to bake all this in.

Reply:
Please not, Rust the language has a bad stance on it and dealing with inconsistent `.so`'s on user machines distributed by distributors instead of us will make it a huge pain. See https://github.com/martinvonz/jj/issues/3575#issuecomment-2088812446 for a good explainer.

Reply:
I think we should manage that by making them Rust feature dependent, which works but makes code as brittle as C++/C #ifdefs if not managed correctly.

Reply:
Rust can load dynamic code fine as long as you define a stable C ABI. You can do this by having a glue layer that matches up a rich Rust API on the host and a rich Rust API in the plugin through a C ABI that the plugin author never has to see. There are existing libraries to accomplish this, including loading plugins with WebAssembly which skips a lot of the OS‐level `dlopen` headache.

Reply:
I'm well aware that Rust can do that, but I still do not like it and Austin has put it nicely in the comment I linked to. It also sounds painful to maintain either way with or without the C ABI and its woes or WASM and its ecosystem.

Reply:
"Dynamic extension loading for forges" is largely a red herring I'm afraid. I'll explain my reasoning briefly but, first off: no, there are no plans to do dynamic linking for anything, much less forge integrations, and I wouldn't read too much into this one particular bullet point on that note, or about this changing anytime soon. Furthermore, unless you specifically mean some other mechanism than the system dynamic linker, I'm going to assume that's what we're talking about.

In the past dynamic linking was useful for a variety of things. People always talk about stuff like "vendoring copies of zlib everywhere" or whatever and security patches, but originally it was actually just a massive time *and* space optimization. It turns out in the 1990s computers were slow and hard drives were slower; if you linked in a copy of a library into two programs, then executed them both, they were slower to start because they had more code pages that needed to be read from disk! Hard drive spindles are slow! Think about the boot process of a Unix machine; it's typically hundreds of short-lived program invocations. It was actually a massive wall clock time optimization to put everything into shared objects; code that would be read from disk 100 times now only needed to be loaded once. I remember reading some ancient hieroglyphics that once told a tale where boot time for a SunOS was slashed from minutes to seconds due to the introduction of shared object linking.

At the same time, in other places like Windows-land, it's also historically been a cultural consideration to allow dynamic loading of DLLs which are popular as a method of extending programs. Extensions are often useful in places because they allow things like:

- Decoupled development velocity: The main program may be aging but the extensions may thrive and add new functionality. The developers are not the same set of people, typically.
- Political/technical differences: some extensions add things that aren't seen as part of the project's focus or even go against the design criteria of the maintainers.
- Different approaches to a desired outcome. It's often useful to have some people explore new approaches.

Dynamic linking also requires:

- Careful technical design: you need a robust ABI/API design up front, because if you just spend time iterating it constantly and breaking plugins, what's the point? There is obviously some give and take here, but broadly speaking it requires a lot of investment in this ABI barrier to make it all work out.
- Cultural considerations: developers *and* extension authors have to actually respect existing ABIs. This is cultural! Not technical! Windows developers and Linux developers treat this completely differently in practice (more on this in a second)
- Particular distribution methods so that things do not explode fantastically because it turns out it's a *really* fiddly and user-unfriendly method of distributing things if it doesn't go just right.

And now I've got the bad news: honestly I feel none of these really apply to us or are so much work as to dwarf the actual tasks we want to accomplish.

- We don't want to decouple development! We're actively thriving. We want to *increase* the velocity of our own development if anything and keep things going. We actively want people to contribute upstream, and it's much easier to get things like deeper code review and shared tooling that way.
- On that note, we do a release every month so we're keen to get things into users hands quickly. So there isn't much overall velocity improvement that can come from having dynamic loading; we could just do releases more frequently too!
- There aren't political or technical difficulties -- everyone wants these features! We all think they're really important, even if we disagree about exact technical procedures to achieve them.
- While there is utility to having different approaches, we have strong opinions about topics like code review and patch submission, so I feel this is an opportunity for us to actually "double down" and really embrace Jujutsu's strengths. For example I don't really know how much different you can actually make a Gerrit integration, when you really get down to it.

And:

- Spending a lot of time designing a robust stable C-ified ABI for all our interfaces is a lot of work! It's tricky! There are platform specific things that suck! Dynamic linking was always a useful but massively complex feature in my experience working in the Haskell world, and it seems similarly "sticky" in Rust. Most importantly, we can spend this time actually improving the Rust codebase more extensively rather than spending time in the weeds.
- I'll be frank, Linux as a community, with a culture, just doesn't give a damn about ABI stability. Windows and even macOS in contrast deeply commit to at least years' worth of ABI/binary stability. In the Linux world, there's a massive amount of cultural inertia which boils down to "just recompile the thing" which completely eliminates the value in having a stable ABI, and isn't a good user experience. The distribution methods are mostly centralized too, so the decoupled development is often not really achieved because your distro has to do the updates for you anyway. We aren't packaged in distros or anything either, for the most part, so largely users just end up compiling things anyway. We shouldn't encourage this, IMO.
- And on the note of distribution methods: well, it's too much to write, but it requires everyone to do the right thing, anytime the wrong thing happens we'll get inscrutible bug reports, and due to the aforementioned bits about recompiling all the time, you often just nerf many benefits to dynamic loading.

I will also finally note two things to close this out:

- **We can still extend Jujutsu, today, without dynamic linking, and without upstream patches**! Just import the crates and add custom commands. Because of this many of the "experimentation" or "go your own way" points above are mitigated. No, it's not using the super cool dynamic linker, just boring Rust compilation. Yes, that has its own downsides. But it's there and largely fills most of the same practical roles. And: we want these features upstream, and at least in my case I'm already an upstream developer &mdash; so it makes sense for PRs like the Gerrit one to go here first. But I could have done all that out of tree, too.
- I like to think of this discussion as a good example of "top down" versus "bottom up" design dynamics, IMO. Dynamic linking is a very "bottom up" concern, it's a concern of building blocks, which is very easy for us to all latch on to (it's a difficult and technically interesting topic.) While "github integration for users" is a very "top-down" concern; it starts from the UX and moves from there. You can't design the UX without a list of features which is where we should start; beginning at dynamic linking doesn't tell us much because it has no relation to the features available! I think top-down design is very important for us as we seek to evolve because it will give us a better chance at a great UX that will make users happier, even if we don't always use some particular technical approach.

This is more than I expected to write, but that's to be expected from me, but I suspect this should probably address most of the talking points people have.

Reply:
Oh, but I will say this: if we *do* want Jujutsu to be extendable in a more open-ended way, which I think is important &mdash; I think a much better "bang for buck" approach is to make the CLI scriptable instead of dynamic loading. I don't have much more to say on this topic for now, but it's something I want to explore sometime in the next year. So maybe keep your ears open (or bug me elsewhere on the topic)

Reply:
Sorry to make you write all of that; I just wanted to address the specific technical point :)

I think that there is value to people being able to add extensions without recompiling all of Jujutsu, in the long run, but I agree that loading code directly, whether via `dlopen` or WebAssembly, is probably not the best approach, and that just adding upstream support for a bunch of forges is the correct approach for most things.

Reply:
Do you think I should make any changes in this doc? It's already unspecified how we're going to implement it and I'm fine with that. Should I explicitly say that it's not yet decided? Should I say that we're likely going to add bundled support for a few popular forges (that's my preference)?

Reply:
I think the latter is a good formulation, since I left a +1 w/o clearly saying which option it was for.

Reply:
I don't feel strongly about dynamic loading specifically, but I do think there'd be value to users if they can just `apt-get install jj; apt-get install some-jj-extension` (or equivalent) and then get updates for those without having to deal with cargo, keeping track of versions, etc. (e.g. I'd hate if a lot of users miss out on critical/security updates because they were forced to compile their own, or if they don't even try out jj because it's too much hassle to get it installed).
As others have pointed out there's many ways to achieve that though.

> Do you think I should make any changes in this doc? 

I'd be fine if no changes are made here.

Reply:
Done.